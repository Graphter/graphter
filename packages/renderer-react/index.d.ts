import ListRenderer from './list-renderer';
export { ListRenderer };
