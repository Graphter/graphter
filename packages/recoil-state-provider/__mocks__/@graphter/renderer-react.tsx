import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)