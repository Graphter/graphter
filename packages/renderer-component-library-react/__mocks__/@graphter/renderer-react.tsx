import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()