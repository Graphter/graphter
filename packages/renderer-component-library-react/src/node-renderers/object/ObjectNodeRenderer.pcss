.objectNodeRenderer{
    display: inline-block;
    padding: 20px;
}
.defaultWrapper{

}
.field-name{

}
.description{
    font-size: 0.7em;
    margin-top: 0;
}