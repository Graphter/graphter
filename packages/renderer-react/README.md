# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.