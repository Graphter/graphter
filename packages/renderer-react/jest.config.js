module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};