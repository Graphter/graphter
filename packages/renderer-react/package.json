{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
