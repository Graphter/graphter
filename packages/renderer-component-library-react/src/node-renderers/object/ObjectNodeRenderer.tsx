import React from "react";
import { NodeConfig, NodeRendererProps } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";
import { createDefault, useNodeData } from "@graphter/renderer-react";
import s from './ObjectNodeRenderer.pcss'

export default function ObjectNodeRenderer(
  {
    config,
    originalNodeData,
    path,
    committed,
    ErrorDisplayComponent
  }: NodeRendererProps
) {
  if (!originalNodeData) originalNodeData = createDefault(config, {})
  useNodeData(path, config, originalNodeData, committed)
  return (
    <div className={s.objectNodeRenderer}>
      {config.children && config.children.map((childConfig, i) => {
        const childRendererRegistration = nodeRendererStore.get(childConfig.type)
        if (!childRendererRegistration) return null
        const ChildTypeRenderer = childRendererRegistration.renderer
        return (
          <DefaultPropertyWrapper config={childConfig} key={childConfig.id}>
            <ChildTypeRenderer
              committed={committed}
              config={childConfig}
              path={[ ...path, childConfig.id ]}
              originalNodeData={originalNodeData[childConfig.id]}
              options={childRendererRegistration.options}
              ErrorDisplayComponent={ErrorDisplayComponent} />
          </DefaultPropertyWrapper>
        )
      })}
    </div>
  );
}

interface DefaultPropertyWrapperProps {
  config: NodeConfig
  children: any
}

function DefaultPropertyWrapper(
  {
    config,
    children
  }: DefaultPropertyWrapperProps
) {
  return (

    <div className={s.defaultWrapper}>
      <label htmlFor={config.id}>{config.name}</label>
      {config.description && <p className={s.description}>{config.description}</p>}
      {children}
    </div>
  )
}