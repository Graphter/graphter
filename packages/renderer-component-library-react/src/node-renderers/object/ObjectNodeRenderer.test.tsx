import React from 'react'
import ObjectNodeRenderer from "./ObjectNodeRenderer";
import { render } from "@testing-library/react";
import * as graphterRenderer from "@graphter/renderer-react";
import clone from "rfdc";
import { when } from "jest-when";

const useNodeDataMock = graphterRenderer.useNodeData as jest.Mock<any>
const nodeRendererStoreGetMock = graphterRenderer.nodeRendererStore.get as jest.Mock<any>
const createDefaultMock = graphterRenderer.createDefault as jest.Mock<any>

describe('<ObjectNodeRenderer />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('render correctly', () => {
    when(nodeRendererStoreGetMock)
      .calledWith('string')
      .mockReturnValue({
        Renderer: () => <div>A child item</div>
      })
    const { queryAllByText } = render(<ObjectNodeRenderer
      originalNodeData={{}}
      originalNodeDataAncestry={[]}
      committed={true}
      config={{
        id: 'author',
        name: 'Author',
        description: 'An author',
        type: 'object',
        children: [
          {
            id: 'name',
            name: 'Name',
            type: 'string'
          },
          {
            id: 'location',
            name: 'Location',
            type: 'string'
          },
        ]
      }}
      configAncestry={[]}
      path={['/']} /> )
    expect(queryAllByText('A child item').length).toBe(2)
  })

  it('should render default data when new', () => {
    createDefaultMock.mockReturnValue({})
    const config = {
      id: 'author',
      name: 'Author',
      description: 'An author',
      type: 'object',
      children: [
        {
          id: 'name',
          name: 'Name',
          type: 'string'
        }
      ]
    }
    render(<ObjectNodeRenderer
      config={clone()(config)}
      configAncestry={[]}
      originalNodeData={undefined}
      originalNodeDataAncestry={[]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(createDefaultMock).toHaveBeenCalledWith(config, {})
    expect(useNodeDataMock).toHaveBeenCalledWith([ '/' ], config, {}, true)
  })

  it('should use the data provider for data', () => {
    const config = {
      id: 'name',
      name: 'Name',
      description: 'The name',
      type: 'list',
      children: [
        {
          id: 'name',
          name: 'Name',
          type: 'string'
        }
      ]
    }
    render(<ObjectNodeRenderer
      config={clone()(config)}
      configAncestry={[]}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      originalNodeDataAncestry={[]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(useNodeDataMock).toHaveBeenCalled()
  })

  it('should render child items with the correct renderers', () => {
    const config = {
      id: 'name',
      name: 'Name',
      description: 'The name',
      type: 'list',
      children: [
        {
          id: 'name',
          name: 'Name',
          type: 'string'
        }
      ]
    }
    render(<ObjectNodeRenderer
      config={clone()(config)}
      configAncestry={[]}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      originalNodeDataAncestry={[]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(nodeRendererStoreGetMock).toHaveBeenCalledWith('string')
  })

  it('should pass the correct props to child renderers', () => {
    const config = {
      id: 'name',
      name: 'Name',
      description: 'The name',
      type: 'list',
      children: [
        {
          id: 'name',
          name: 'Name',
          type: 'string'
        }
      ]
    }
    const rendererMock = jest.fn().mockReturnValue(<div>The child renderer</div>)
    when(nodeRendererStoreGetMock)
      .calledWith('string')
      .mockReturnValueOnce({
        Renderer: rendererMock
      })
    const { queryByText } = render(<ObjectNodeRenderer
      config={clone()(config)}
      configAncestry={[]}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      originalNodeDataAncestry={[]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(queryByText('The child renderer')).not.toBeNull()
    expect(rendererMock).toHaveBeenCalled()
    const props = rendererMock.mock.calls[0][0]
    expect(props).toMatchSnapshot()
  })

})

export {}