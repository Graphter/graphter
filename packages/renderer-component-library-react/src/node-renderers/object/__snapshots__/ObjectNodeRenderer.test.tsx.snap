// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ObjectNodeRenderer /> should pass the correct props to child renderers 1`] = `
Object {
  "ErrorDisplayComponent": undefined,
  "committed": true,
  "config": Object {
    "id": "name",
    "name": "Name",
    "type": "string",
  },
  "configAncestry": Array [
    Object {
      "children": Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "string",
        },
      ],
      "description": "The name",
      "id": "name",
      "name": "Name",
      "type": "list",
    },
  ],
  "options": undefined,
  "originalNodeData": undefined,
  "originalNodeDataAncestry": Array [
    Array [
      "item 1 value",
      "item 2 value",
    ],
  ],
  "path": Array [
    "/",
    "name",
  ],
}
`;
